### PR TITLE
refactor: tests in `CompilerSuite`

### PR DESCRIPTION
**Pull request by ninaandrup**
Related to #3144

### CODE REVIEW

File: main/test/flix/CompilerSuite.scala [hunk 1]
```diff
@@ -5,6 +5,8 @@ import ca.uwaterloo.flix.util.{FlixSuite, Options}
 class CompilerSuite extends FlixSuite(incremental = true) {
   implicit val options: Options = Options.TestWithLibAll
 
+  mkTestDir("main/test/flix/")
```
Comment (ninaandrup) [inline main/test/flix/CompilerSuite.scala:8]: @magnus-madsen This is not working. The problem is, that in `main/test/flix/` there is `Test.Def.Generalization.flix` and `Test.Derives.Hash.flix` that were not included in this files before. Should they be deleted or refactored?

Comment (magnus-madsen) [reply main/test/flix/CompilerSuite.scala]: See if you can make them work

File: main/test/flix/Test.Def.Generalization.flix [hunk 2]
```diff
@@ -77,8 +75,8 @@ namespace Test/Def/Generalization {
     @test
     def testGen25(): #{A(Int32), B(Int32), C(Int32) | r} = #{A(21). B(42).}
 
-    @test
-    def testGen26(): #{A(Int32) | r1} -> #{B(Int32) | r2} -> #{A(Int32), B(Int32) | r} = x -> y -> x <+> y
+    //@test
+    //def testGen26(): #{A(Int32) | r1} -> #{B(Int32) | r2} -> #{A(Int32), B(Int32) | r} = x -> y -> x <+> y
```
Comment (ninaandrup) [inline main/test/flix/Test.Def.Generalization.flix:79]: VScode gives me the error
```
>> The type scheme: '∀(?row6200511). Unit -> (#{ ?row6200511 } -> (#{ ?row6200511 } -> #{ ?row6200511 }))' cannot be generalized to '∀(r1, r2, r). Unit -> (#{ A(Int32) | r1 } -> (#{ B(Int32) | r2 } -> #{ A(Int32), B(Int32) | r }))'.

79 |     def testGen26(): #{A(Int32) | r1} -> #{B(Int32) | r2} -> #{A(Int32), B(Int32) | r} = x -> y -> x <+> y
             ^^^^^^^^^
             unable to generalize the type scheme.

The declared type does not match the inferred type:

  Declared: ∀(r1, r2, r). Unit -> (#{ A(Int32) | r1 } -> (#{ B(Int32) | r2 } -> #{ A(Int32), B(Int32) | r }))
  Inferred: ∀(?row6200511). Unit -> (#{ ?row6200511 } -> (#{ ?row6200511 } -> #{ ?row6200511 }))

Explanation:

The declared type variables:
  r1 of kind: 'SchemaRow'.
  r2 of kind: 'SchemaRow'.
  r of kind: 'SchemaRow'.

The inferred type variables:
  ?row6200511 of kind: 'SchemaRow'.
```

File: main/test/flix/Test.Def.Generalization.flix [hunk 3]
```diff
@@ -159,17 +157,17 @@ namespace Test/Def/Generalization {
...
+    def testLeq28(): (Unit -> b & (e1 and e2 and e3)) -> b & (e3 and e2 and e1) = f -> f()
 
-    @test
+    /*@test
     def testLeq29(x: Choice[String, true, true]): Int32 =
         choose x {
             case Absent     => 1
             case Present(_) => 2
-        }
+        }*/
```
Comment (ninaandrup) [inline main/test/flix/Test.Def.Generalization.flix:170]: this compiles in VScode, but when I run it, I get the error
```
Exception in thread "main" dev.flix.runtime.MatchError: Non-exhaustive match
```

**Review by ninaandrup — COMMENTED** [review]
I'm almost done with this, but there are some tests in `Test.Def.Generalization.flix` that I cannot seem to make work.

**Comment by magnus-madsen** [comment]
@mlutze Could you look at some of these issues? Are the tests just outdated?

File: main/test/flix/Test.Def.Generalization.flix [hunk 4]
```diff
@@ -42,19 +40,19 @@ namespace Test/Def/Generalization {
...
     @test
     def testGen15(): (a -> a & e) -> (a -> a & e) = x -> x
 
     @test
     def testGen16(): a -> (b -> a) = x -> (_ -> x)
 
-    @test
-    def testGen17(): r -> {x :: Int32 | r} = r -> {+x = 21 | r}
+    //@test
+    //def testGen17(): r -> {x :: Int32 | r} = r -> {+x = 21 | r}
```
[the remark this answers is not in the document]
Comment (mlutze) [reply main/test/flix/Test.Def.Generalization.flix:52]: We need to wrap the `r` record row in a record constructor to turn it into a type:

```scala
def testGen17(): { | r} -> {x :: Int32 | r} = r -> {+x = 21 | r}
```

File: main/test/flix/Test.Def.Generalization.flix [hunk 5]
```diff
@@ -42,19 +40,19 @@ namespace Test/Def/Generalization {
...
 
-    @test
-    def testGen18(): {x :: Int32 | r} -> r = r -> { -x | r}
+    //@test
+    //def testGen18(): {x :: Int32 | r} -> r = r -> { -x | r}
```
[the remark this answers is not in the document]
Comment (mlutze) [reply main/test/flix/Test.Def.Generalization.flix:55]: As above
```scala
def testGen18(): {x :: Int32 | r} -> { | r} = r -> { -x | r}
```

**Comment by mlutze** [comment]
> @mlutze Could you look at some of these issues? Are the tests just outdated?

I was able to help with the Kind issues. I'm not sure about the constraint set issue. The type ascription definitely seems wrong, but I don't know how to type that in a nontrivial way. And the choice issue seems like a bug. I'll take a deeper look into that tomorrow.

**Comment by ninaandrup** [comment]
Tests are failing, because `mkTestDir` does not work with [this line](https://github.com/flix/flix/blob/f27d31c6d965ccdf2d17a24198bbb602c8f473ab/main/test/flix/CompilerSuite.scala#L146). Should I will undo the `mkTestDir` to merge this PR? Then I can do the `mkTestDir` in another PR?

**Comment by magnus-madsen** [comment]
> Tests are failing, because `mkTestDir` does not work with [this line](https://github.com/flix/flix/blob/f27d31c6d965ccdf2d17a24198bbb602c8f473ab/main/test/flix/CompilerSuite.scala#L146). Should I will undo the `mkTestDir` to merge this PR? Then I can do the `mkTestDir` in another PR?

Yes, open another PR with the change to mkTestDir.

@mlutze The time has come.

**Comment by mlutze** [comment]
> > Tests are failing, because `mkTestDir` does not work with [this line](https://github.com/flix/flix/blob/f27d31c6d965ccdf2d17a24198bbb602c8f473ab/main/test/flix/CompilerSuite.scala#L146). Should I will undo the `mkTestDir` to merge this PR? Then I can do the `mkTestDir` in another PR?
> 
> Yes, open another PR with the change to mkTestDir.
> 
> @mlutze The time has come.

Should I beg for forgiveness?

**Comment by magnus-madsen** [comment]
@ninaandrup After this is done, are we done with all the duplicate code? Or is there still some remaining?

**Comment by ninaandrup** [comment]
> @ninaandrup After this is done, are we done with all the duplicate code? Or is there still some remaining?

This should get rid of all the duplicate code from all the files in `"main/test/flix/"`

**Comment by ninaandrup** [comment]
> > Tests are failing, because `mkTestDir` does not work with [this line](https://github.com/flix/flix/blob/f27d31c6d965ccdf2d17a24198bbb602c8f473ab/main/test/flix/CompilerSuite.scala#L146). Should I will undo the `mkTestDir` to merge this PR? Then I can do the `mkTestDir` in another PR?
> 
> Yes, open another PR with the change to mkTestDir.
> 
> @mlutze The time has come.

New PR: https://github.com/flix/flix/pull/3787

**Comment by magnus-madsen** [comment]
Very nice! I am very happy to have this in because now all these tests use the Flix library itself :)

**Pull request merged**